zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/X509ZNodeGroupAclProvider.java

```diff
@@ -0,0 +1,180 @@
+  }
+
+  public X509ZNodeGroupAclProvider(X509TrustManager trustManager, X509KeyManager keyManager) {
+    this.trustManager = trustManager;
+    this.keyManager = keyManager;
+  }
+
+  @Override
+  public KeeperException.Code handleAuthentication(ServerObjs serverObjs, byte[] authData) {
+    ServerCnxn cnxn = serverObjs.getCnxn();
+    X509Certificate clientCert;
+    try {
+      clientCert = X509AuthenticationUtil.getAuthenticatedClientCert(cnxn, trustManager);
+    } catch (KeeperException.AuthFailedException e) {
+      return KeeperException.Code.AUTHFAILED;
+    }
+
+    // The clientId can be any string matched and extracted using regex from Subject Distinguished
+    // Name or Subject Alternative Name from x509 certificate.
+    // The clientId string should be an URI for client and map the client to certain domain.
+    //The user can use the properties defined in X509AuthenticationUtil to extract a desired string as clientId.
+    String clientId;
+    try {
+      clientId = X509AuthenticationUtil.getClientId(clientCert);
+    } catch (Exception e) {
+      // Failed to extract clientId from certificate
+      LOG.error(logStrPrefix + "Failed to extract URI from certificate for session 0x{}",
+          Long.toHexString(cnxn.getSessionId()), e);
+      return KeeperException.Code.OK;
+    }
+
+    // User belongs to super user group
+    if (clientId.equals(System.getProperty(ZOOKEEPER_ZNODEGROUPACL_SUPERUSER))) {
+      cnxn.addAuthInfo(new Id("super", clientId));
+      LOG.info("Authenticated Id '{}' as super user", clientId);
+      return KeeperException.Code.OK;
+    }
+
+    // Get authorized domain names for client
+    Set<String> domains = getUriDomainMappingHelper(serverObjs.getZks()).getDomains(clientId);
+    if (domains.isEmpty()) {
+      // If no domain name is found, use URI as domain name
+      domains = new HashSet<>();
+      domains.add(clientId);
+    }
+
+    Set<String> superUserDomainNames =
+        ZNodeGroupAclProperties.getInstance().getSuperUserDomainNames();
+    for (String domain : domains) {
+      // Grant super user privilege to users belong to super user domains
+      if (superUserDomainNames.contains(domain)) {
+        cnxn.addAuthInfo(new Id("super", clientId));
+        LOG.info(
+            logStrPrefix + "Id '{}' belongs to superUser domain '{}', authenticated as super user",
+            clientId, domain);
+      } else {
+        cnxn.addAuthInfo(new Id(getScheme(), domain));
+        LOG.info(logStrPrefix + "Authenticated Id '{}' for Scheme '{}', Domain '{}'.", clientId,
+            getScheme(), domain);
+      }
+    }
+
+    return KeeperException.Code.OK;
+  }
+
+  @Override
+  public boolean matches(ServerObjs serverObjs, MatchValues matchValues) {
+    for (Id id : serverObjs.getCnxn().getAuthInfo()) {
+      if (id.getId().equals(matchValues.getAclExpr()) || id.getId()
+          .equals(System.getProperty(ZOOKEEPER_ZNODEGROUPACL_SUPERUSER))) {
+        return true;
```

Comment:
Curious, we're already giving `super` in `handleAuthentication()`. Why is this doing a super user check again? Could you please clarify why this is necessary?

Reply:
We have discussed about a break-glass mechanism and decided to keep an extra way to authorize super user, remember? Let me know if you think we need more discussion around this.

Reply:
Discussed offline. The super user check here is too make this matched() logic completed. But the same check is already covered in checkAcl() method in ZooKeeperServer.class, and since right now checkAcl() is the only place that matched() is called, we don't need to do a double check here. Left a comment in the code in case later the matched() being called in some other place, then we should add this super user check back.